test/units/testsuite-73.sh

```diff
@@ -0,0 +1,404 @@
+        output=$(systemctl show-environment)
+        assert_in "LANG=$i" "$output"
+        assert_not_in "LC_CTYPE=" "$output"
+    done
+
+    # test if localed auto-runs locale-gen
+    if command -v locale-gen >/dev/null 2>&1 &&
+           ! localectl list-locales | grep -F "de_DE.UTF-8"; then
+
+        # clear previous locale
+        systemctl stop systemd-localed.service
+        rm -f /etc/locale.conf /etc/default/locale
+
+        # change locale
+        assert_rc 0 localectl set-locale de_DE.UTF-8
+        if [[ -f /etc/default/locale ]]; then
+            assert_eq "$(cat /etc/default/locale)" "LANG=de_DE.UTF-8"
+        else
+            assert_eq "$(cat /etc/locale.conf)" "LANG=de_DE.UTF-8"
+        fi
+        assert_in "System Locale: LANG=de_DE.UTF-8" "$(localectl)"
+        assert_in "LANG=de_DE.UTF-8" "$(systemctl show-environment)"
+
+        # ensure tested locale exists and works now
+        assert_in "de_DE.UTF-8" "$(localectl list-locales)"
+    fi
+
+    # reset locale to original
+    restore_locale
+    trap - EXIT
+}
+
+backup_keymap() {
+    if [[ -f /etc/vconsole.conf ]]; then
+        cp /etc/vconsole.conf /tmp/vconsole.conf.bak
+    fi
+
+    if [[ -f /etc/X11/xorg.conf.d/00-keyboard.conf ]]; then
+        cp /etc/X11/xorg.conf.d/00-keyboard.conf /tmp/00-keyboard.conf.bak
+    fi
+
+    # Debian/Ubuntu specific file
+    if [[ -f /etc/default/keyboard ]]; then
+        cp /etc/default/keyboard /tmp/default-keyboard.bak
+    fi
+
+    mkdir -p /etc/default
+}
+
+restore_keymap() {
+    if [[ -f /tmp/vconsole.conf.bak ]]; then
+        mv /tmp/vconsole.conf.bak /etc/vconsole.conf
+    else
+        rm -f /etc/vconsole.conf
+    fi
+
+    if [[ -f /tmp/00-keyboard.conf.bak ]]; then
+        mv /tmp/00-keyboard.conf.bak /etc/X11/xorg.conf.d/00-keyboard.conf
+    else
+        rm -f /etc/X11/xorg.conf.d/00-keyboard.conf
+    fi
+
+    if [[ -f /tmp/default-keyboard.bak ]]; then
+        mv /tmp/default-keyboard.bak /etc/default/keyboard
+    else
+        rm -f /etc/default/keyboard
+        rmdir --ignore-fail-on-non-empty /etc/default
+    fi
+}
+
+wait_vconsole_setup() {
+    local i ss
+    for ((i=0;i<20;i++)); do
+        if (( i != 0 )); then sleep .5; fi
+        ss="$(systemctl --property SubState --value show systemd-vconsole-setup.service)"
+        if [[ "$ss" == "exited" || "$ss" == "dead" || "$ss" == "condition" ]]; then
+            return 0
+        elif [[ "$ss" == "failed" ]]; then
+            echo "WARNING: systemd-vconsole-setup.service failed, ignoring." >&2
+            systemctl reset-failed systemd-vconsole-setup.service
+            return 0
+        fi
+    done
+
+    systemctl status systemd-vconsole-setup.service
+    return 1
+}
+
+test_vc_keymap() {
+    local i output
+
+    if [[ -z "$(localectl list-keymaps)" ]]; then
+        echo "No vconsole keymap installed, skipping test."
+        return
+    fi
+
+    backup_keymap
+    trap restore_keymap EXIT
```

Comment:
This could be, theoretically, simplified by using

```suggestion
    trap restore_keymap RETURN
```

and dropping all the

```bash
    restore_keymap
    trap - EXIT
```

statements below. Similarly in other functions.